staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go

```diff
@@ -117,12 +190,223 @@ func (a *ApplySet) addLabels(objects []*resource.Info) error {
+	}
+	if idLabel != a.ID() {
+		return fmt.Errorf("ApplySet parent object %q exists and has incorrect value for label %q (got: %s, want: %s)", a.ParentRef(), ApplySetParentIDLabel, idLabel, a.ID())
+	}
+
+	if a.currentResources, err = parseResourcesAnnotation(annotations, a.restMapper); err != nil {
+		// TODO: handle GVRs for now-deleted CRDs
+		return fmt.Errorf("parsing ApplySet annotation on %q: %w", a.ParentRef(), err)
+	}
+	a.currentNamespaces = parseNamespacesAnnotation(annotations)
+	if a.parent.IsNamespaced() {
+		a.currentNamespaces.Insert(a.parent.Namespace)
+	}
+	return nil
+}
+
+func getLabelsAndAnnotations(obj runtime.Object) (map[string]string, map[string]string) {
+	accessor := meta.NewAccessor()
+	annotations := make(map[string]string)
+	labels := make(map[string]string)
+
+	if data, err := accessor.Annotations(obj); err == nil {
+		annotations = data
+	}
+	if data, err := accessor.Labels(obj); err == nil {
+		labels = data
+	}
+	return labels, annotations
+}
+
+func toolingBaseName(toolAnnotation string) string {
+	parts := strings.Split(toolAnnotation, "/")
+	if len(parts) >= 2 {
+		return strings.Join(parts[:len(parts)-1], "/")
+	}
+	return toolAnnotation
+}
+
+func parseResourcesAnnotation(annotations map[string]string, mapper meta.RESTMapper) (map[schema.GroupVersionResource]*meta.RESTMapping, error) {
+	annotation, ok := annotations[ApplySetGRsAnnotation]
+	if !ok {
+		// The spec does not require this annotation. However, 'missing' means 'perform discovery' (as opposed to 'present but empty', which means ' this is an empty set').
+		// We return an error because we do not currently support dynamic discovery in kubectl apply.
+		return nil, fmt.Errorf("kubectl requires the %q annotation to be set on all ApplySet parent objects", ApplySetGRsAnnotation)
+	}
+	mappings := make(map[schema.GroupVersionResource]*meta.RESTMapping)
+	for _, grString := range strings.Split(annotation, ",") {
+		gr := schema.ParseGroupResource(grString)
+		gvk, err := mapper.KindFor(gr.WithVersion(""))
+		if err != nil {
+			return nil, fmt.Errorf("invalid group resource in %q annotation: %w", ApplySetGRsAnnotation, err)
+		}
+		mapping, err := mapper.RESTMapping(gvk.GroupKind())
+		if err != nil {
+			return nil, fmt.Errorf("could not find kind for resource in %q annotation: %w", ApplySetGRsAnnotation, err)
+		}
+		mappings[mapping.Resource] = mapping
+	}
+	return mappings, nil
+}
+
+func parseNamespacesAnnotation(annotations map[string]string) sets.Set[string] {
+	namespaces := make(sets.Set[string])
+	annotation, ok := annotations[ApplySetAdditionalNamespacesAnnotation]
+	if !ok { // this annotation is completely optional
+		return namespaces
+	}
+	for _, ns := range strings.Split(annotation, ",") {
+		namespaces.Insert(ns)
+	}
+	return namespaces
+}
+
+type ApplySetUpdateMode string
+
+var SetUpdateSuccessful ApplySetUpdateMode = "latest"
```

Comment:
Nit: Maybe "BeforeApplyPrune" and "AfterApplyPrune" might be clearer, or `ExpandingBeforeApply` and `ShrinkingAfterPrune` but IDK..  Not user-facing so doesn't matter.